File: examples/langchain/studio/graph.py
import os
from datetime import datetime

from configuration import AgentConfigurable
from langchain_arcade import ArcadeToolManager
from langchain_core.messages import HumanMessage, ToolMessage
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from langgraph.graph import END, START, MessagesState, StateGraph
from langgraph.prebuilt import ToolNode

# Initialize the Arcade Tool Manager with your API key
arcade_api_key = os.getenv("ARCADE_API_KEY")
openai_api_key = os.getenv("OPENAI_API_KEY")

toolkit = ArcadeToolManager(api_key=arcade_api_key)
# Retrieve tools compatible with LangGraph
tools = toolkit.get_tools(langgraph=True)
tool_node = ToolNode(tools)

PROMPT_TEMPLATE = f"""
You are a helpful assistant who can use tools to help users with tasks
Today's date is {datetime.now().strftime("%Y-%m-%d")}

ALL RESPONSES should be in plain text and not markdown.
"""
# prompt for the main agent
prompt = ChatPromptTemplate.from_messages([
    ("system", PROMPT_TEMPLATE),
    ("placeholder", "{messages}"),
])
# Initialize the language model with your OpenAI API key
model = ChatOpenAI(model="gpt-4o", api_key=openai_api_key).bind_tools(tools)
prompted_model = prompt | model


def call_agent(state):
    """Define the agent function that invokes the model"""
    messages = state["messages"]
    # replace placeholder with messages from state
    response = prompted_model.invoke({"messages": messages})
    return {"messages": [response]}


def should_continue(state: MessagesState, config: dict):
    """Function to determine the next step based on the model's response"""
    last_message = state["messages"][-1]
    if last_message.tool_calls:
        user_id = config["configurable"].get("user_id")
        tool_name = state["messages"][-1].tool_calls[0]["name"]
        auth_response = toolkit.authorize(tool_name, user_id)
        if auth_response.status == "completed":
            return "tools"
        else:
            # If the tool requires authorization, proceed to the authorization step
            return "authorization"
    # If no tool calls are present, end the workflow
    return END


def wait_for_auth(state: MessagesState):
    last_message = state["messages"][-1]
    if isinstance(last_message, HumanMessage):
        return "agent"
    return "tools"


def authorize(state: MessagesState, config: dict):
    """Function to handle tool authorization"""
    user_id = config["configurable"].get("user_id")
    tool_name = state["messages"][-1].tool_calls[0]["name"]
    auth_response = toolkit.authorize(tool_name, user_id)

    auth_message = (
        f"Please authorize the application in your browser:\n\n {auth_response.authorization_url}"
    )
    tool_call_id = state["messages"][-1].tool_calls[0]["id"]
    response = ToolMessage(
        content=auth_message,
        tool_call_id=tool_call_id,
    )
    # Add the new message to the message history and add a new human message
    # saying that the agent should try again
    try_message = HumanMessage(
        content="Please try the previous tool call again now that you are authorized."
    )
    return {"messages": [response, try_message]}


# Build the workflow graph
workflow = StateGraph(MessagesState, AgentConfigurable)

# Add nodes to the graph
workflow.add_node("agent", call_agent)
workflow.add_node("tools", tool_node)
workflow.add_node("authorization", authorize)
# workflow.add_node("wait_for_auth", wait_for_auth)

# Define the edges and control flow
workflow.add_edge(START, "agent")
workflow.add_conditional_edges("agent", should_continue, ["authorization", "tools", END])
workflow.add_edge("authorization", "agent")
workflow.add_edge("tools", "agent")

# Compile the graph with an interrupt after the authorization node
# so that we can prompt the user to authorize the application
graph = workflow.compile(interrupt_after=["authorization"])

File: examples/langgraph-ts/src/graph.ts
import { Arcade } from "@arcadeai/arcadejs";
import { executeOrAuthorizeZodTool, toZod } from "@arcadeai/arcadejs/lib";
import type { AIMessage } from "@langchain/core/messages";
import type { RunnableConfig } from "@langchain/core/runnables";
import { type Tool, tool } from "@langchain/core/tools";
import { MessagesAnnotation, StateGraph } from "@langchain/langgraph";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import { ChatOpenAI } from "@langchain/openai";
import { ConfigurationSchema, ensureConfiguration } from "./configuration.ts";

// Initialize Arcade
const arcade = new Arcade();

// Replace this with your application's user ID (e.g. email address, UUID, etc.)
const USER_ID = "user@example.com";
// Get the Arcade tools, you can customize the toolkit (e.g. "github", "notion", "google", etc.)
const googleToolkit = await arcade.tools.list({ toolkit: "google", limit: 30 });

/**
 * LangGraph requires tools to be defined using Zod, a TypeScript-first schema validation library
 * that has become the standard for runtime type checking. Zod is particularly valuable because it:
 * - Provides runtime type safety and validation
 * - Offers excellent TypeScript integration with automatic type inference
 * - Has a simple, declarative API for defining schemas
 * - Is widely adopted in the TypeScript ecosystem
 *
 * Arcade provides `toZod` to convert our tools into Zod format, making them compatible
 * with LangGraph.
 *
 * The `executeOrAuthorizeZodTool` helper function simplifies authorization.
 * It checks if the tool requires authorization: if so, it returns an authorization URL,
 * otherwise, it runs the tool directly without extra boilerplate.
 *
 * Learn more: https://docs.arcade.dev/home/use-tools/get-tool-definitions#get-zod-tool-definitions
 */
const arcadeTools = toZod({
	tools: googleToolkit.items,
	client: arcade,
	userId: USER_ID,
	executeFactory: executeOrAuthorizeZodTool, // Checks if tool is authorized and executes it, or returns authorization URL if needed
});

// Convert Arcade tools to LangChain tools
const tools = arcadeTools.map(({ name, description, execute, parameters }) =>
	tool(execute, {
		name,
		description,
		schema: parameters,
	}),
);

// Define the function that calls the model
async function callModel(
	state: typeof MessagesAnnotation.State,
	config: RunnableConfig,
): Promise<typeof MessagesAnnotation.Update> {
	/** Call the LLM powering our agent. **/
	const configuration = ensureConfiguration(config);

	/**
	 * Initialize the model and bind the tools
	 */
	const model = new ChatOpenAI({
		model: configuration.model,
		apiKey: process.env.OPENAI_API_KEY,
	}).bindTools(tools);

	const response = await model.invoke([
		{
			role: "system",
			content: configuration.systemPromptTemplate.replace(
				"{system_time}",
				new Date().toISOString(),
			),
		},
		...state.messages,
	]);

	// We return a list, because this will get added to the existing list
	return { messages: [response] };
}

// Define the function that determines whether to continue or not
function routeModelOutput(state: typeof MessagesAnnotation.State): string {
	const messages = state.messages;
	const lastMessage = messages[messages.length - 1];
	// If the LLM is invoking tools, route there.
	if ((lastMessage as AIMessage)?.tool_calls?.length) {
		return "tools";
	}
	// Otherwise end the graph.

	return "__end__";
}

// Define a new graph. We use the prebuilt MessagesAnnotation to define state:
// https://langchain-ai.github.io/langgraphjs/concepts/low_level/#messagesannotation
const workflow = new StateGraph(MessagesAnnotation, ConfigurationSchema)
	// Define the two nodes we will cycle between
	.addNode("callModel", callModel)
	.addNode("tools", new ToolNode(tools))
	// Set the entrypoint as `callModel`
	// This means that this node is the first one called
	.addEdge("__start__", "callModel")
	.addConditionalEdges(
		// First, we define the edges' source node. We use `callModel`.
		// This means these are the edges taken after the `callModel` node is called.
		"callModel",
		// Next, we pass in the function that will determine the sink node(s), which
		// will be called after the source node is called.
		routeModelOutput,
	)
	// This means that after `tools` is called, `callModel` node is called next.
	.addEdge("tools", "callModel");

// Finally, we compile it!
// This compiles it into a graph you can invoke and deploy.
export const graph = workflow.compile({
	interruptBefore: [], // if you want to update the state before calling the tools
	interruptAfter: [],
});
